selendroid-server/src/main/java/io/selendroid/server/handler/Actions.java

```diff
@@ -64,51 +43,49 @@ public Response safeHandle(HttpRequest request) throws JSONException {
     SelendroidLogger.info("Got actions request");
     JSONArray payload = getPayload(request).getJSONArray("payload");
     int actionChainCount = payload.length();
-    int longestActionChain = longestActionChain(payload);
+    boolean stillRunning = true;
+    boolean allPaused;
 
-    // Create contexts for all chains of actions
-    List<ActionContext> chainContexts = new ArrayList<ActionContext>();
+    List<ActionChain> actionChains = new ArrayList<ActionChain>();
     for (int i = 0; i < actionChainCount; i++) {
-      chainContexts.add(new ActionContext());
+      actionChains.add(new ActionChain(payload.getJSONObject(i)));
     }
 
-    /*
-       Action chains are a sequence of actions for a single finger on the touch screen device
-       These chains are executed in parallel when only one finger wont suffice,
-       chains can be of different lengths so we iterate over the longest one -
-       padding out shorter chains with wait commands once they are done.
-    */
-    for (int i = 0; i < longestActionChain; i++) {
-      for (int j = 0; j < actionChainCount; j++) {
-        JSONObject actionChain = payload.getJSONObject(j);
-        String inputDevice = actionChain.getString("inputDevice");
-
-        JSONArray actions = actionChain.getJSONArray("actions");
-        if (actions.length() > i) {
-          ActionContext context = chainContexts.get(j);
-          JSONObject action = actions.getJSONObject(i);
+    while (stillRunning) {
+      stillRunning = false;
+      allPaused = true;
+      for (ActionChain chain : actionChains) {
+        if (chain.isPaused()) {
+          stillRunning = true;
+        }
+        else if (chain.hasNext()) {
+          stillRunning = true;
+          JSONObject action = chain.next();
           String actionName = action.getString("name");
 
-          SelendroidLogger.info("Performing action " + inputDevice + "/" + actionName);
+          SelendroidLogger.info("Performing action " + chain.getInputDevice() + "/" + actionName);
 
-          // Pauses are now handled separately as we only wait for the longest pause.
           if (!actionName.equals(TouchActionName.PAUSE)) {
-            ActionHandler handler = ActionHandler.getHandlerForInputDevice(inputDevice);
-            handler.handle(actionName, getSelendroidDriver(request), action, context);
+            allPaused = false;
+            ActionHandler handler = ActionHandler.getHandlerForInputDevice(chain.getInputDevice());
+            handler.handle(actionName, getSelendroidDriver(request), action, chain.getContext());
           }
         }
       }
 
-      int pauseTime = longestPause(i, payload);
-      if (pauseTime > 0) {
+      // Sleep the thread until the next non-pause action
+      if(allPaused) {
+        int pauseTime = shortestPause(actionChains);
```

Comment:
so, this shouldn't be the shortest pause, we should keep the original 'longest' pause. The intention of multi-action chains is to gate each 'tick' with all the other actions at it's tick.

    [
      [down, pause 10, release],
      [pause 0, pause 1, tap]
    ]

This should not result in the tap being performed 9 seconds before the release. This example payload should perform the "down" command, then a 10 second pause (the longest) in the 'tick', then release and tap at the "same" time (the release would get called first, then the tap)

Also... both original and as proposed are flawed with finding what the sleep interval should be. It should be determined in the current 'tick' of the action chain, not searching the entire payload (or actionChains) and using that time.